### PR TITLE
Updated fix for issue 42 ( Add maps.me mwm export )

### PR DESCRIPTION
**Pull request by unithiyashree**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -68,7 +68,7 @@ media/*
 static
 !ui/static
 migrations
-!jobs/migrations
+#!jobs/migrations
```
Comment (unithiyashree) [inline .gitignore]: No. for adding a migration file I had to do that. reverting it back.

File: core/settings/dev.py [hunk 2]
```diff
@@ -133,7 +133,7 @@
             'propagate': True,
             'level': 'DEBUG',
         },
-        'hot_exports': {
+        'core': {
```
Comment (unithiyashree) [inline core/settings/dev.py]: Nope. So, should I make a seperate PR for this ?

File: utils/pbf.py [hunk 3]
```diff
@@ -69,5 +69,5 @@ def convert(self, ):
     debug = False
     if config.get('debug'):
         debug = True
-    o2p = OSMToPBF(osm=osm, pbf=pbf, debug=debug)
+    o2p = OSMToPBF(osm=osm, pbffile=pbf, debug=debug)
```
Comment (unithiyashree) [inline utils/pbf.py]: No.

File: .gitignore (hunk 1 above)
Comment (pnorman) [inline .gitignore]: You've commented out this line. Should it be deleted instead?

File: README.md [hunk 4]
```diff
@@ -197,14 +211,14 @@ they are pushed to a Celery Worker for processing. At least two celery workers n
 
 From a 'hotosm' virtualenv directory (use screen), run:
 
-<code>export DJANGO_SETTINGS_MODULE=core.settings.your_settings_module</code>
+<code>export DJANGO_SETTINGS_MODULE=core.settings.your_settings_module</code>
```
Comment (pnorman) [inline README.md]: There's extra whitespace here at the end of the line.

File: README.md [hunk 5]
```diff
@@ -197,14 +211,14 @@ they are pushed to a Celery Worker for processing. At least two celery workers n
...
-<code>export DJANGO_SETTINGS_MODULE=core.settings.your_settings_module</code>
+<code>export DJANGO_SETTINGS_MODULE=core.settings.your_settings_module</code>  
 
 <code>$ celery -A core worker --loglevel debug --logfile=celery.log</code>.
 
 This will start a celery worker which will process export tasks. An additional celery worker needs to be started to handle purging of expired unpublished
 export jobs. From another hotosm virtualenv terminal session in the project top-level directory, run:
 
-<code>export DJANGO_SETTINGS_MODULE=core.settings.your_settings_module</code>
+<code>export DJANGO_SETTINGS_MODULE=core.settings.your_settings_module</code>
```
Comment (pnorman) [inline README.md]: ditto on whitespace

File: core/settings/project.py [hunk 6]
```diff
@@ -36,11 +37,14 @@
...
 
+# the root url for maps me binary files
+EXPORT_MWM_ROOT = ''
+
 # location of the garmin config file
 GARMIN_CONFIG = '/home/ubuntu/www/hotosm/utils/conf/garmin_config.xml'
 
 # url to overpass api endpoint
-OVERPASS_API_URL = 'http://localhost/interpreter'
+OVERPASS_API_URL = 'http://overpass-api.de/api/interpreter'
```
Comment (pnorman) [inline core/settings/project.py]: This needs fixing

File: core/settings/dev.py (hunk 2 above)
Comment (pnorman) [inline core/settings/dev.py]: Is this change related to MWM?

**Comment by pnorman** [comment]
Replaces #101, for #42.

File: tasks/export_tasks.py [hunk 7]
```diff
@@ -194,6 +194,24 @@ def run(self, run_uid=None, stage_dir=None, job_name=None):
...
+    Returns the path to tha mwm file.
+    """
+    name = 'MWM Export'
+    abort_on_error = True
+    def run(self, run_uid=None, stage_dir=None, job_name=None):
+        self.update_task_state(run_uid=run_uid, name=self.name)
+        osm = stage_dir + job_name + '.osm'
+        pbffile = stage_dir + job_name + '.pbf'
+        mwmfile = stage_dir + job_name + '.mwm'
+        logger.debug("fsddfsdfsa")
```
Comment (pnorman) [inline tasks/export_tasks.py]: leftover debugging statement?

File: utils/pbf.py (hunk 3 above)
Comment (pnorman) [inline utils/pbf.py]: Did this not work before?

File: utils/mwm.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    def __init__(self, osm=None, pbffile=None, mwmfile=None, debug=False):
+        self.osm = osm
+        if not os.path.exists(self.osm):
+            raise IOError('Cannot find raw OSM data for this task.')
+        self.pbffile = pbffile
+        self.mwmfile = mwmfile
+        if not self.pbffile:
+            # create pbf path from osm path.
+            root = self.osm.split('.')[0]
```
Comment (Zverik) [inline utils/mwm.py]: I'd suggest `self.osm.rsplit('.', 1)[0]` so filenames with multiple dots get processed correctly.

File: utils/mwm.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            self.pbffile = root + '.pbf' 
+        if not self.mwmfile:
+            # create mwm path from osm path.
+            root = self.osm.split('.')[0]
+            self.mwmfile = root + '.mwm' 
+        self.debug = debug
+        mwm_root_path = settings.EXPORT_MWM_ROOT
+        self.cmd = Template('osmconvert $osm --out-pbf >$pbf')
```
Comment (Zverik) [inline utils/mwm.py]: Aren't pbf files generated somewhere in the exporting process already? I guess you can specify this dependency and use a pre-generated pbf.

Comment (Zverik) [inline utils/mwm.py]: Also, [this pull-request](https://github.com/mapsme/omim/pull/487) adds `.osm` source support for generate_mwm.sh.

File: utils/mwm.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        if (returncode != 0):
+            logger.error('%s', stderr)
+            raise Exception, "osmconvert failed with return code: {0}".format(returncode)
+        if(self.debug):
+            print 'Osmconvert returned: %s' % returncode
+        convert_cmd_mwm = self.cmd_mwm.safe_substitute({'pbf': self.pbffile})
+        if(self.debug):
+            print 'Running: %s' % convert_cmd_mwm
+        proc_mwm = subprocess.Popen(convert_cmd_mwm, shell=True, executable='/bin/bash', stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        (stdout, stderr) = proc_mwm.communicate()
```
Comment (Zverik) [inline utils/mwm.py]: You can use `(_, stderr) = ...` if you are not using `stdout` variable.

File: utils/mwm.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        returncode = proc_mwm.wait()
```
Comment (Zverik) [inline utils/mwm.py]: `returncode = proc.returncode` or just use the latter in `if` later, since `wait()` call is indirectly called by `communicate()`.

File: utils/mwm.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        if (returncode != 0):
+            logger.error('%s', stderr)
+            raise Exception, "generate_mwm failed with return code: {0}".format(returncode)
+        if(self.debug):
+            print 'generate_mwm returned: %s' % returncode
```
Comment (Zverik) [inline utils/mwm.py]: Please stick to a single format notation, preferably with `'...returned: {0}'.format(returncode)`

Comment (unithiyashree) [inline utils/mwm.py]: Ok Sure :) will update.

File: utils/mwm.py (hunk 11 above)
Comment (unithiyashree) [inline utils/mwm.py]: Ok.

File: utils/mwm.py (hunk 10 above)
Comment (unithiyashree) [inline utils/mwm.py]: Ok will update . thanks

File: utils/mwm.py (hunk 9 above)
Comment (unithiyashree) [inline utils/mwm.py]: So, using this we can directly convert from osm to mwm ?

Comment (Zverik) [inline utils/mwm.py]: Yes, if the source file has `.osm` extension.

**Comment by Zverik** [comment]
Needs to be tested by somebody with a dev environment, but looks good to me.

**Comment by unithiyashree** [comment]
Yes.. :+1:  @dodobas can you test it ?

**Comment by unithiyashree** [comment]
can you please test this?

**Comment by bjohare** [comment]
@unithiyashree working on this at the moment. Will let you know when I've finished testing.

**Comment by unithiyashree** [comment]
hi @bjohare.. Have you tested the PR?

**Comment by mojodna** [comment]
@unithiyashree I just took a stab at testing and merging, but ran into problems building `omim` (https://github.com/mapsme/omim/issues/5515) on 14.04.

**Comment by mojodna** [comment]
The QT 5.5 PPA has changed to `ppa:beineri/opt-qt551-trusty`

**Comment by unithiyashree** [comment]
@mojodna  Thank you for looking in to it.I contributed long back and the whole setup isn't with me. Let me know whether it is working or not , if not let me know the problem.

**Comment by mojodna** [comment]
@unithiyashree it's not working right now. the root cause (or next problem) appears to be https://github.com/mapsme/omim/issues/5515 (and I don't know what to use as an alternative PPA).

**Comment by mojodna** [comment]
We updated our base image to Ubuntu 16.04, so dependency problems have changed. qt55 (from the referenced PPA) isn't available for 16.04--will a newer version suffice? I was also having problems with compilers / Boost-related warnings, so knowing which compiler version / packages (for 16.04) would be super helpful.

I realized that we (probably?) only need the generator tool, so prepending with `CONFIG=gtool` is important.

@Zverik can you help answer these questions?

**Comment by Zverik** [comment]
I have built the gnerator tool on Ubuntu 16.04 just now. All I had to install were packages `qt5-default` and `libsqlite3-dev`. I built it with `CONFIG="gtool no-tests" tools/unix/build_omim.sh -cr`

**Comment by mojodna** [comment]
@Zverik fantastic, thanks!! I'll take another whack at this soon.

**Comment by mojodna** [comment]
@Zverik I'm getting lots of errors complaining about `auto_ptr`s, so I don't think I have the right build toolchain installed:

```
In file included from /usr/local/src/omim/3party/boost/boost/smart_ptr/shared_ptr.hpp:28:0,
                 from /usr/local/src/omim/3party/boost/boost/smart_ptr/shared_array.hpp:24,
                 from /usr/local/src/omim/3party/boost/boost/shared_array.hpp:17,
                 from /usr/local/src/omim/std/shared_array.hpp:7,
                 from /usr/local/src/omim/coding/reader.hpp:8,
                 from /usr/local/src/omim/coding/multilang_utf8_string.hpp:3,
                 from /usr/local/src/omim/indexer/feature_meta.hpp:3,
                 from /usr/local/src/omim/indexer/mwm_set.hpp:11,
                 from /usr/local/src/omim/indexer/feature_decl.hpp:3,
                 from /usr/local/src/omim/ugc/types.hpp:3,
                 from /usr/local/src/omim/ugc/storage.hpp:3,
                 from /usr/local/src/omim/ugc/api.hpp:5,
                 from /usr/local/src/omim/ugc/api.cpp:1:
/usr/local/src/omim/3party/boost/boost/smart_ptr/detail/shared_count.hpp:396:33: warning: 'template<class> class std::auto_ptr' is deprecated [-Wdeprecated-declarations]
     explicit shared_count( std::auto_ptr<Y> & r ): pi_( new sp_counted_impl_p<Y>( r.get() ) )
...
At global scope:
cc1plus: warning: unrecognized command line option '-Wno-unused-local-typedef'
```

```bash
$ dpkg -l | grep llvm
ii  libllvm3.8:amd64                1:3.8-2ubuntu4                     amd64        Modular compiler and toolchain technologies, runtime library
$ dpkg -l | grep gcc
ii  gcc                             4:5.3.1-1ubuntu1                   amd64        GNU C compiler
ii  gcc-5                           5.4.0-6ubuntu1~16.04.4             amd64        GNU C compiler
ii  gcc-5-base:amd64                5.4.0-6ubuntu1~16.04.4             amd64        GCC, the GNU Compiler Collection (base package)
ii  gcc-6-base:amd64                6.0.1-0ubuntu1                     amd64        GCC, the GNU Compiler Collection (base package)
ii  libgcc-5-dev:amd64              5.4.0-6ubuntu1~16.04.4             amd64        GCC support library (development files)
ii  libgcc1:amd64                   1:6.0.1-0ubuntu1                   amd64        GCC support library
ii  qtchooser                       52-gae5eeef-2build1~gcc5.2         amd64        Wrapper to select between Qt development binary versions
```

**Comment by mojodna** [comment]
```bash
$ dpkg -l | grep libc
ii  libc-bin                        2.23-0ubuntu9                      amd64        GNU C Library: Binaries
ii  libc-dev-bin                    2.23-0ubuntu9                      amd64        GNU C Library: Development binaries
ii  libc6:amd64                     2.23-0ubuntu9                      amd64        GNU C Library: Shared libraries
ii  libc6-dev:amd64                 2.23-0ubuntu9                      amd64        GNU C Library: Development Libraries and Header Files
ii  libcc1-0:amd64                  5.4.0-6ubuntu1~16.04.4             amd64        GCC cc1 plugin for GDB
ii  linux-libc-dev:amd64            4.4.0-83.106                       amd64        Linux Kernel Headers for development

$ dpkg -l | grep libstdc
ii  libstdc++-5-dev:amd64           5.4.0-6ubuntu1~16.04.4             amd64        GNU Standard C++ Library v3 (development files)
ii  libstdc++6:amd64                5.4.0-6ubuntu1~16.04.4             amd64        GNU Standard C++ Library v3
```

**Comment by mojodna** [comment]
I can also point you at the packages we do install: https://github.com/hotosm/osm-export-tool2/blob/master/Dockerfile#L9-L42

**Comment by Zverik** [comment]
I have the same packages. Could you do `grep -A 1 'error:' omim-build-release/build_error.log`

**Comment by mojodna** [comment]
```
/usr/local/src/omim/routing/routing_session.cpp:15:48: fatal error: 3party/Alohalytics/src/alohalytics.h: No such file or directory
```

**Comment by mojodna** [comment]
I forgot to init submodules. Trying again.

**Comment by mojodna** [comment]
Success!
